README.md
# aes_crypt 

## Introduction

aes_crypt is a library for Dart and Flutter developers that uses 256-bit AES algorithm 
to encrypt/decrypt files, plain text and binary data. It is fully compatible with the 
[AES Crypt](https://www.aescrypt.com/) file format.
It can be used to integrate AES Crypt functionality into your own Dart or Flutter applications.
All algorithms are implemented in pure Dart and work in all platforms.

aes_crypt writes and reads version 2 (latest) of the AES Crypt file specification. Backwards compatibility 
with reading the version 1 is implemented but untested. 
Output .aes files are fully compatible with any software using the AES Crypt standard file format. 
This library is accompanied by clients and libraries for different operating systems
and programming languages.
For more information about AES Crypt and AES Crypt compatible 
applications for other platforms, please visit [AESCrypt's official website](https://www.aescrypt.com).  
 
## Features

- 256-bit AES encryption format.
- File-to-file encryption and decryption.
- Memory-to-file encryption, file-to-memory decryption.
- Password can be in Unicode (like "密碼 パスワード пароль كلمة السر").
- Support for asynchronous file system reading/writing.
- Encrypted files have .aes extension which clients on other operating systems recognize.
- Compatible software available for Windows, Linux, Mac OS, Android and iOS 
- now typesafe with null safety
(https://www.aescrypt.com/download/).

## Usage

In the `pubspec.yaml` of your project, add the following dependency:
```yaml
dependencies:
  ...
  aes_crypt: ^0.1.1
```

In your Dart code add the following import:
```dart
import 'dart:typed_data';
import 'package:aes_crypt/aes_crypt.dart';
```

Initialization:
```dart
var crypt = AesCrypt('my cool password');
```
or
```dart
var crypt = AesCrypt();
crypt.setPassword('my cool password');
```

Optionally you can set overwrite mode for the file write operations:
```dart
// Overwrites the file if it exists.
crypt.setOverwriteMode(AesCryptOwMode.on);

// If the file exists, stops the operation and throws 'AesCryptException'
// exception with 'AesCryptExceptionType.destFileExists' type (see 
// example1.dart in 'example'  folder). This mode is set by default.
crypt.setOverwriteMode(AesCryptOwMode.warn);

// If the file exists, adds index '(1)' to its' name and tries to save. 
// If such file also exists, adds '(2)' to its name, then '(3)', etc. 
crypt.setOverwriteMode(AesCryptOwMode.rename);
```

*Notice: All functions having 'Sync' at the end of their names are synchronous.
If you need asynchronous ones, please just remove 'Sync' from the end of function name.*


File encryption/decryption:
```dart
// Encrypts the file srcfile.txt and saves encrypted file under original name 
// with '.aes' extention added (srcfile.txt.aes). You can specify relative or 
// direct path to it. To save the file into current directory specify it 
// either as './srcfile.txt' or as 'srcfile.txt'.
crypt.encryptFileSync('srcfile.txt');

// Encrypts the file srcfile.txt and saves encrypted file under 
// the name enc_file.txt.aes
crypt.encryptFileSync('srcfile.txt', 'enc_file.txt.aes');

// Decrypts the file srcfile.txt.aes and saves decrypted file under 
// the name srcfile.txt
crypt.decryptFileSync('srcfile.txt.aes');

// Decrypts the file srcfile.txt.aes and saves decrypted file under 
// the name dec_file.txt
crypt.decryptFileSync('srcfile.txt.aes', 'dec_file.txt');
```

Text <=> file encryption/decryption:
```dart
String decryptedText;

// Plain text to be encrypted
String srcText = 'some text';

// Encrypts the text as UTF8 string and saves it into 'mytext.txt.aes' file.
crypt.encryptTextToFileSync(srcText, 'mytext.txt.aes');
// Encrypts the text as UTF16 Big Endian string and saves it 
// into 'mytext.txt.aes' file.
crypt.encryptTextToFileSync(srcText, 'mytext.txt.aes', utf16: true);
// Encrypts the text as UTF16 Little Endian string and saves it 
// into 'mytext.txt.aes' file.
crypt.encryptTextToFileSync(srcText, 'mytext.txt.aes', utf16: true, endian: Endian.little);
// Add 'bom: true' as an argument if you want to add byte order mark 
// at the beginning of the text string before the encryption. For example:
// crypt.encryptTextToFileSync(srcText, 'mytext.txt.aes', bom: true);

// Decrypts the file and interprets it based on byte order mark if it has one.
// Otherwise it will be interpreted as UTF8 text.
decryptedString = crypt.decryptTextFromFileSync('mytext.txt.aes');
// Decrypts the file and interprets it based on byte order mark if it has one.
// Otherwise it will be interpreted as UTF16 Big Endian text.
decryptedString = crypt.decryptTextFromFileSync('mytext.txt.aes', utf16: true);
// Decrypts the file and interprets it based on byte order mark if it has one.
// Otherwise it will be interpreted as UTF16 Little Endian text.
decryptedString = crypt.decryptTextFromFileSync('mytext.txt.aes', utf16: true, endian: Endian.little);

```

Binary data <=> file encryption/decryption:
```dart
// Binary data to be encrypted
Uint8List srcData = Uint8List.fromList([1,2,3,4,5]);

// Encrypts the data and saves it into mydata.bin.aes file.
crypt.encryptDataToFileSync(srcData, 'mydata.bin.aes');

// Decrypt the data from 'mydata.bin.aes' file
Uint8List decryptedData = crypt.decryptDataFromFileSync('mydata.bin.aes');
```

Binary data AES encryption/decryption:
```dart
// The encryption key. It should be 128, 192 or 256 bits long.
Uint8List key = Uint8List.fromList([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16]);

// The initialization vector used in advanced cipher modes. 
// It must be 128 bits long.
Uint8List iv = Uint8List.fromList([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16]);

// AES mode of operation. It can be one of the next values:
//    AesMode.ecb - ECB (Electronic Code Book)
//    AesMode.cbc - CBC (Cipher Block Chaining)
//    AesMode.cfb - CFB (Cipher Feedback)
//    AesMode.ofb - OFB (Output Feedback)
// By default the mode is AesMode.cbc
AesMode mode = AesMode.cbc; // Ok. I know it's meaningless here.

// Sets the encryption key and IV.
crypt.aesSetKeys(key, iv);
// Sets cipher mode
crypt.aesSetMode(mode);

// If you wish you can set the key, IV and cipher mode in one function.
//crypt.aesSetParams(key, iv, mode);

// The binary data to be encrypted
Uint8List srcData = Uint8List.fromList([1,2,3,4,5]);

// Encrypts the data. Padding scheme - null byte (0x00).
Uint8List encryptedData = crypt.aesEncrypt(srcData);
// Decrypts the data
Uint8List decryptedData = crypt.aesDecrypt(encryptedData);
```

SHA256 and HMAC-SHA256 computation:
```dart
// The source data
Uint8List srcData = Uint8List.fromList([1,2,3,4,5,6,7,8,9]);

// Computes SHA256 hash
Uint8List hash = crypt.sha256(srcData);

// Secret cryptographic key for HMAC
Uint8List key = Uint8List.fromList([1,2,3]);

// Computes HMAC-SHA256 code
Uint8List hmac = crypt.hmacSha256(key, srcData);
```


## Future plans

- reducing the memory usage for large file processing
- asynchronous encrypting/decrypting
- support for streams
- support for key files

## Support

Please file feature requests and bugs at the [issue tracker](https://github.com/alexgoussev/aes_crypt/issues)

I would be grateful if you help me to correct grammar and syntactical errors 
for this readme and library documentation.  


## Acknowledgments

- to Phil Nicholls for [PHP-AES-File-Encryption library](https://github.com/philios33/PHP-AES-File-Encryption) 
- to IgoAtM for [PHP aesCrypt complient class](https://forums.packetizer.com/viewtopic.php?f=72&t=403)
- to author of [phpAES](http://www.phpaes.com) library ([github](https://github.com/phillipsdata/phpaes))
- to authors of [phpseclib](http://phpseclib.sourceforge.net/) ([github](https://github.com/phpseclib/phpseclib))
- to Steven Roose for [pointycastle](https://pub.dev/packages/pointycastle) Dart library ([github](https://github.com/PointyCastle/pointycastle))


CHANGELOG.md
## 0.1.0

- Initial version

## 0.1.1

- Improve internal code structure. Move some code to separate classes.

## 0.1.2

- Added Null Safety


test/aes_crypt_test.dart
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:aes_crypt/aes_crypt.dart';
import 'package:test/test.dart';

void main() {
  var random = Random();

  group('Algorithms', () {
    AesCrypt crypt = AesCrypt();
    Uint8List iv =
        Uint8List.fromList(List<int>.generate(16, (i) => random.nextInt(256)));
    Uint8List key =
        Uint8List.fromList(List<int>.generate(32, (i) => random.nextInt(256)));
    crypt.aesSetKeys(key, iv);
    int srcDataLen = 100016;
    Uint8List srcData = Uint8List.fromList(
        List<int>.generate(srcDataLen, (i) => random.nextInt(256)));

    test('Test AES CBC encryption/decryption', () {
      crypt.aesSetMode(AesMode.cbc);
      Uint8List encData = crypt.aesEncrypt(srcData);
      Uint8List decData = crypt.aesDecrypt(encData);
      expect(srcData.isEqual(decData), equals(true));
    });

    test('Test AES ECB encryption/decryption', () {
      crypt.aesSetMode(AesMode.ecb);
      Uint8List encData = crypt.aesEncrypt(srcData);
      Uint8List decData = crypt.aesDecrypt(encData);
      expect(srcData.isEqual(decData), equals(true));
    });

    test('Test AES CFB encryption/decryption', () {
      crypt.aesSetMode(AesMode.cfb);
      Uint8List encData = crypt.aesEncrypt(srcData);
      Uint8List decData = crypt.aesDecrypt(encData);
      expect(srcData.isEqual(decData), equals(true));
    });

    test('Test AES OFB encryption/decryption', () {
      crypt.aesSetMode(AesMode.ofb);
      Uint8List encData = crypt.aesEncrypt(srcData);
      Uint8List decData = crypt.aesDecrypt(encData);
      expect(srcData.isEqual(decData), equals(true));
    });
  });

  AesCrypt crypt = AesCrypt();
  crypt.setPassword('passw 密碼 パスワード пароль كلمة السر ꜩꝕ 𐌰𐍉 𝕬𝖃');
  crypt.setOverwriteMode(AesCryptOwMode.warn);

  int srcDataLen = 100003;
  Uint8List srcData = Uint8List.fromList(
      List<int>.generate(srcDataLen, (i) => random.nextInt(256)));

  group('Encryption/decryption', () {
    test('Test `encryptFileSync()` and `decryptFileSync()` functions', () {
      String src_filepath = './test/testfile.txt';
      String enc_filepath = './test/testfile2.txt.aes';
      String source_data1 = File(src_filepath).readAsStringSync();
      enc_filepath = crypt.encryptFileSync(src_filepath, enc_filepath);
      String dec_filepath = crypt.decryptFileSync(enc_filepath);
      File(enc_filepath).deleteSync();
      String source_data2 = File(dec_filepath).readAsStringSync();
      File(dec_filepath).deleteSync();
      expect(source_data2, equals(source_data1));
    });

    test('Test `encryptFile()` and `decryptFile()` functions', () async {
      String src_filepath = './test/testfile.txt';
      String enc_filepath = './test/testfile2.txt.aes';
      String source_data1 = await File(src_filepath).readAsString();
      enc_filepath = await crypt.encryptFile(src_filepath, enc_filepath);
      String dec_filepath = await crypt.decryptFile(enc_filepath);
      await File(enc_filepath).delete();
      String source_data2 = await File(dec_filepath).readAsString();
      await File(dec_filepath).delete();
      expect(source_data2, equals(source_data1));
    });

    test(
        'Test `decryptFileSync()` functions on a file encypted by AES Crypt software',
        () {
      String src_filepath = './test/testfile.txt';
      String enc_filepath = './test/testfile.txt.aes';
      String dec_filepath = './test/testfile2.txt';
      String source_data1 = File(src_filepath).readAsStringSync();
      dec_filepath = crypt.decryptFileSync(enc_filepath, dec_filepath);
      String source_data2 = File(dec_filepath).readAsStringSync();
      File(dec_filepath).deleteSync();
      expect(source_data2, equals(source_data1));
    });

    String enc_filepath = './test/testfile2.txt.aes';

    test(
        'Test `encryptDataToFileSync()` and `decryptDataFromFileSync()` functions',
        () {
      enc_filepath = crypt.encryptDataToFileSync(srcData, enc_filepath);
      Uint8List decrypted_data = crypt.decryptDataFromFileSync(enc_filepath);
      File(enc_filepath).deleteSync();
      expect(srcData.isEqual(decrypted_data), equals(true));
    });

    test('Test `encryptDataToFile()` and `decryptDataFromFile()` functions',
        () async {
      enc_filepath = await crypt.encryptDataToFile(srcData, enc_filepath);
      Uint8List decrypted_data = await crypt.decryptDataFromFile(enc_filepath);
      await File(enc_filepath).delete();
      expect(srcData.isEqual(decrypted_data), equals(true));
    });

    String decString;
    String srcString =
        'hglakj 密碼 パスワード фбмгцз كلمة السر ꜩꝕ 𐌰𐍉 𝕬𝖃 aalkjhflaeiuwoefdnscvsmnskdjfhoweqirhowqefasdnl';
    String encFilepath = './test/testfile2.txt.aes';

    test('Encrypt/decrypt UTF8 string <=> file', () {
      crypt.encryptTextToFileSync(srcString, encFilepath); // bom = false
      decString = crypt.decryptTextFromFileSync(encFilepath);
      File(encFilepath).delete();
      expect(decString, equals(srcString));
    });

    test('Encrypt/decrypt UTF8 string with BOM <=> file', () {
      crypt.encryptTextToFileSync(srcString, encFilepath,
          bom: true); // bom = true
      decString = crypt.decryptTextFromFileSync(encFilepath);
      File(encFilepath).delete();
      expect(decString, equals(srcString));
    });

    test('Encrypt/decrypt UTF16 BE string <=> file', () {
      crypt.encryptTextToFileSync(srcString, encFilepath,
          utf16: true); // bom = false, endian = Endian.big
      decString = crypt.decryptTextFromFileSync(encFilepath, utf16: true);
      File(encFilepath).delete();
      expect(decString, equals(srcString));
    });

    test('Encrypt/decrypt UTF16 BE string with BOM <=> file', () {
      crypt.encryptTextToFileSync(srcString, encFilepath,
          utf16: true, bom: true); // bom = true, endian = Endian.big
      decString = crypt.decryptTextFromFileSync(encFilepath);
      File(encFilepath).delete();
      expect(decString, equals(srcString));
    });

    test('Encrypt/decrypt UTF16 LE string <=> file', () {
      crypt.encryptTextToFileSync(srcString, encFilepath,
          utf16: true,
          endian: Endian.little); // bom = false, endian = Endian.little
      decString = crypt.decryptTextFromFileSync(encFilepath,
          utf16: true, endian: Endian.little);
      File(encFilepath).delete();
      expect(decString, equals(srcString));
    });

    test('Encrypt/decrypt UTF16 LE string with BOM <=> file', () {
      crypt.encryptTextToFileSync(srcString, encFilepath,
          utf16: true,
          endian: Endian.little,
          bom: true); // bom = true, endian = Endian.little
      decString = crypt.decryptTextFromFileSync(encFilepath);
      File(encFilepath).delete();
      expect(decString, equals(srcString));
    });

//    test('', () {
//      expect(true, equals(true));
//    });
  });
}

extension _Uint8ListExtension on Uint8List {
  bool isEqual(Uint8List other) {
    if (identical(this, other)) return true;
    if (this != null && other == null) return false;
    int length = this.length;
    if (length != other.length) return false;
    for (int i = 0; i < length; i++) {
      if (this[i] != other[i]) return false;
    }
    return true;
  }
}


lib/src/exceptions.dart
part of aes_crypt2;

/// Enum that specifies the type of [AesCryptException] exception.
enum AesCryptExceptionType {
  /// Exception of this type is thrown when [AesCryptOwMode] is set as
  /// [AesCryptOwMode.warn] and some encryption or decryption function
  /// tries to overwrite existing file.
  destFileExists,
}

/// Exception thrown when some error is happened. A type of the error is indicated
/// by [type] class member.
///
/// At present there is only one type: [AesCryptExceptionType.destFileExists].
/// AesCryptException having AesCryptExceptionType.destFileExists type is thrown
/// when overwrite mode set as [AesCryptOwMode.warn] and there is a file with the same name
/// as the file which is going to be written during encryption or decryption process.
class AesCryptException implements Exception {
  /// Message describing the problem.
  final String message;

  /// Type of an exception.
  final AesCryptExceptionType type;

  /// Creates a new AesCryptException with an error message [message] and type [type].
  const AesCryptException(this.message, this.type);

  /// Returns a string representation of this object.
  @override
  String toString() => message;
}

/// Error thrown when a function is passed an unacceptable argument.
class AesCryptArgumentError extends ArgumentError {
  /// Creates a new AesCryptArgumentError with an error message [message]
  /// describing the erroneous argument.
  AesCryptArgumentError(String message) : super(message);

  /// Throws [AesCryptArgumentError] if [argument] is: null [Object], empty [String] or empty [Iterable]
  static void checkNullOrEmpty(Object? argument, String message) {
    if (argument == null ||
        ((argument is String) ? argument.isEmpty : false) ||
        ((argument is Iterable) ? argument.isEmpty : false)) {
      throw AesCryptArgumentError(message);
    }
  }
}

/// Exception thrown when the file system operation fails.
class AesCryptFsException extends FileSystemException {
  /// Creates a new AesCryptFsException with an error message [message],
  /// optional file system path [path] and optional OS error [osError].
  const AesCryptFsException(String message,
      [String? path = '', OSError? osError])
      : super(message, path, osError);
}

/// Exception thrown when an integrity of encrypted data is compromised.
class AesCryptDataException implements Exception {
  /// Message describing the problem.
  final String message;

  /// Creates a new AesCryptDataException with an error message [message].
  const AesCryptDataException(this.message);

  /// Returns a string representation of this object.
  @override
  String toString() => message;
}


lib/src/aescrypt.dart
part of aes_crypt2;

/// Enum that specifies the overwrite mode for write file operations
/// during encryption or decryption process.
enum AesCryptOwMode {
  /// If the file exists, stops the operation and throws [AesCryptException]
  /// exception with [AesCryptExceptionType.destFileExists] type.
  /// This mode is set by default.
  warn,

  /// If the file exists, adds index '(1)' to its' name and tries to save.
  /// If such file also exists, adds '(2)' to its name, then '(3)', etc.
  rename,

  /// Overwrites the file if it exists.
  on,
}

/// Enum that specifies the mode of operation of the AES algorithm.
enum AesMode {
  /// ECB (Electronic Code Book)
  ecb,

  /// CBC (Cipher Block Chaining)
  cbc,

  /// CFB (Cipher Feedback)
  cfb,

  /// OFB (Output Feedback)
  ofb,
}

/// Wraps encryption and decryption methods and algorithms.
class AesCrypt {
  final _aes = _Aes();

  String? _password;
  Uint8List? _passBytes;
  AesCryptOwMode? _owMode;
  Map<String, List<int>>? _userdata;

  /// Creates the library wrapper.
  ///
  /// Optionally sets encryption/decryption password as [password].
  AesCrypt([String password = '']) {
    _password = password;
    _passBytes = password.toUtf16Bytes(Endian.little);
    _owMode = AesCryptOwMode.warn;
    setUserData();
  }

  /// Sets encryption/decryption password.
  void setPassword(String password) {
    AesCryptArgumentError.checkNullOrEmpty(password, 'Empty password.');
    _password = password;
    _passBytes = password.toUtf16Bytes(Endian.little);
  }

  /// Sets overwrite mode [mode] for write file operations during encryption
  /// or decryption process.
  ///
  /// Available modes:
  ///
  /// [AesCryptOwMode.warn] - If the file exists, stops the operation and
  /// throws [AesCryptException] exception with
  /// [AesCryptExceptionType.destFileExists] type. This mode is set by default.
  ///
  /// [AesCryptOwMode.rename] - If the file exists, adds index '(1)' to its' name
  /// and tries to save. If such file also exists, adds '(2)' to its name, then '(3)', etc.
  ///
  /// [AesCryptOwMode.on] - Overwrite the file if it exists.
  void setOverwriteMode(AesCryptOwMode mode) => _owMode = mode;

  /// Sets standard extension tags used in the AES Crypt file format.
  ///
  /// Extension tags available:
  ///
  /// [createdBy] is a developer-defined text string that identifies the software
  /// product, manufacturer, or other useful information (such as software version).
  ///
  /// [createdOn] indicates the date that the file was created.
  /// The format of the date string is YYYY-MM-DD.
  ///
  /// [createdAt] indicates the time that the file was created. The format of the date string
  /// is in 24-hour format like HH:MM:SS (e.g, 21:15:04). The time zone is UTC.
  void setUserData(
      {String createdBy = 'Dart aes_crypt library',
      String createdOn = '',
      String createdAt = ''}) {
    String key;
    _userdata = {};
    if (createdBy.isNotEmpty) {
      key = 'CREATED_BY';
      _userdata![key] = createdBy.toUtf8Bytes();
      if (key.length + _userdata![key]!.length + 1 > 255) {
        throw AesCryptArgumentError(
            'User data \'$key\' is too long. Total length should not exceed 255 bytes.');
      }
    }
    if (createdOn.isNotEmpty) {
      key = 'CREATED_DATE';
      _userdata![key] = createdOn.toUtf8Bytes();
      if (key.length + _userdata![key]!.length + 1 > 255) {
        throw AesCryptArgumentError(
            'User data \'$key\' is too long. Total length should not exceed 255 bytes.');
      }
    }
    if (createdAt.isNotEmpty) {
      key = 'CREATED_TIME';
      _userdata![key] = createdAt.toUtf8Bytes();
      if (key.length + _userdata![key]!.length + 1 > 255) {
        throw AesCryptArgumentError(
            'User data \'$key\' is too long. Total length should not exceed 255 bytes.');
      }
    }
  }

  /// Encrypts binary data [srcData] to [destFilePath] file synchronously.
  ///
  /// Returns [String] object containing the path to encrypted file.
  String encryptDataToFileSync(Uint8List srcData, String destFilePath) {
    destFilePath = destFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        destFilePath, 'Empty encrypted file path.');
    return _Cryptor.init(_passBytes, _owMode, _userdata)
        .encryptDataToFileSync(srcData, destFilePath);
  }

  /// Encrypts a plain text string [srcString] to [destFilePath] file synchronously.
  ///
  /// By default the text string will be converted to a list of UTF-8 bytes before
  /// it is encrypted.
  ///
  /// If the argument [utf16] is set to [true], the text string will be converted
  /// to a list of UTF-16 bytes. Endianness depends on [endian] argument.
  ///
  /// If the argument [bom] is set to [true], BOM (Byte Order Mark) is appended
  /// to the beginning of the text string before it is encrypted.
  ///
  /// Returns [String] object containing the path to encrypted file.
  String encryptTextToFileSync(String srcString, String destFilePath,
      {bool utf16 = false, Endian endian = Endian.big, bool bom = false}) {
    Uint8List bytes = utf16
        ? srcString.toUtf16Bytes(endian, bom)
        : srcString.toUtf8Bytes(bom) as Uint8List;
    return encryptDataToFileSync(bytes, destFilePath);
  }

  /// Encrypts binary data [srcData] to [destFilePath] file asynchronously.
  ///
  /// Returns [Future<String>] that completes with the path to encrypted file
  /// once the entire operation has completed.
  Future<String> encryptDataToFile(
      Uint8List srcData, String destFilePath) async {
    destFilePath = destFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        destFilePath, 'Empty encrypted file path.');
    return await _Cryptor.init(_passBytes, _owMode, _userdata)
        .encryptDataToFile(srcData, destFilePath);
  }

  /// Encrypts a plain text string [srcString] to [destFilePath] file asynchronously.
  ///
  /// By default the text string will be converted to a list of UTF-8 bytes before
  /// it is encrypted.
  ///
  /// If the argument [utf16] is set to [true], the text string will be converted
  /// to a list of UTF-16 bytes. Endianness depends on [endian] argument.
  ///
  /// If the argument [bom] is set to [true], BOM (Byte Order Mark) is appended
  /// to the beginning of the string before it is encrypted.
  ///
  /// Returns [Future<String>] that completes with the path to encrypted file
  /// once the entire operation has completed.
  Future<String> encryptTextToFile(String srcString, String destFilePath,
      {bool utf16 = false,
      Endian endian = Endian.big,
      bool bom = false}) async {
    Uint8List bytes = utf16
        ? srcString.toUtf16Bytes(endian, bom)
        : srcString.toUtf8Bytes(bom) as Uint8List;
    return await encryptDataToFile(bytes, destFilePath);
  }

  /// Encrypts [srcFilePath] file to [destFilePath] file synchronously.
  ///
  /// If the argument [destFilePath] is not specified, encrypted file name is created
  /// as a concatenation of [srcFilePath] and '.aes' file extension.
  ///
  /// If encrypted file exists, the behaviour depends on [AesCryptOwMode].
  ///
  /// Returns [String] object containing the path to encrypted file.
  String encryptFileSync(String srcFilePath, [String destFilePath = '']) {
    srcFilePath = srcFilePath.trim();
    destFilePath = destFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        srcFilePath, 'Empty source file path.');
    if (srcFilePath == destFilePath) {
      throw AesCryptArgumentError(
          'Source file path and encrypted file path are the same.');
    }
    return _Cryptor.init(_passBytes, _owMode, _userdata)
        .encryptFileSync(srcFilePath, destFilePath);
  }

  /// Encrypts [srcFilePath] file to [destFilePath] file asynchronously.
  ///
  /// If the argument [destFilePath] is not specified, encrypted file name is created
  /// as a concatenation of [srcFilePath] and '.aes' file extension.
  ///
  /// If encrypted file exists, the behaviour depends on [AesCryptOwMode].
  ///
  /// Returns [Future<String>] that completes with the path to encrypted file
  /// once the entire operation has completed.
  Future<String> encryptFile(String srcFilePath,
      [String destFilePath = '']) async {
    srcFilePath = srcFilePath.trim();
    destFilePath = destFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        srcFilePath, 'Empty source file path.');
    if (srcFilePath == destFilePath) {
      throw AesCryptArgumentError(
          'Source file path and encrypted file path are the same.');
    }
    return await _Cryptor.init(_passBytes, _owMode, _userdata)
        .encryptFile(srcFilePath, destFilePath);
  }

  /// Decrypts binary data from [srcFilePath] file synchronously.
  ///
  /// Returns [Uint8List] object containing decrypted data.
  Uint8List decryptDataFromFileSync(String srcFilePath) {
    srcFilePath = srcFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        srcFilePath, 'Empty source file path.');
    return _Cryptor.init(_passBytes, _owMode, _userdata)
        .decryptDataFromFileSync(srcFilePath);
  }

  /// Decrypts a plain text from [srcFilePath] file synchronously.
  ///
  /// If BOM (Byte Order Mark) is present in decrypted data, interprets the data
  /// in accordance with BOM. Otherwise the interpretation will depend on [utf16]
  /// and [endian] arguments.
  ///
  /// If the argument [utf16] is set to [true], decrypted data will be interpreted
  /// as a list of UTF-16 bytes. Endianness depends on [endian] argument.
  /// Otherwise the data will be interpreted as a list of UTF-8 bytes.
  ///
  /// Returns [String] object containing decrypted text.
  String decryptTextFromFileSync(String srcFilePath,
      {bool utf16 = false, Endian endian = Endian.big}) {
    Uint8List decData = decryptDataFromFileSync(srcFilePath);
    String srcString;
    if ((decData[0] == 0xFE && decData[1] == 0xFF) ||
        (decData[0] == 0xFF && decData[1] == 0xFE)) {
      srcString = decData.toUtf16String();
    } else if (decData[0] == 0xEF && decData[1] == 0xBB && decData[2] == 0xBF) {
      srcString = decData.toUtf8String();
    } else {
      srcString =
          utf16 ? decData.toUtf16String(endian) : decData.toUtf8String();
    }
    return srcString;
  }

  /// Decrypts binary data from [srcFilePath] file asynchronously.
  ///
  /// Returns [Future<Uint8List>] that completes with decrypted data
  /// once the entire operation has completed.
  Future<Uint8List> decryptDataFromFile(String srcFilePath) async {
    srcFilePath = srcFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        srcFilePath, 'Empty source file path.');
    return await _Cryptor.init(_passBytes, _owMode, _userdata)
        .decryptDataFromFile(srcFilePath);
  }

  /// Decrypts a plain text from [srcFilePath] file asynchronously.
  ///
  /// If BOM (Byte Order Mark) is present in decrypted data, interprets the data
  /// in accordance with BOM. Otherwise the interpretation will depend on [utf16]
  /// and [endian] arguments.
  ///
  /// If the argument [utf16] is set to [true], decrypted data will be interpreted
  /// as a list of UTF-16 bytes. Endianness depends on [endian] argument.
  /// Otherwise the data will be interpreted as a list of UTF-8 bytes.
  ///
  /// Returns [Future<String>] that completes with decrypted text
  /// once the entire operation has completed.
  Future<String> decryptTextFromFile(String srcFilePath,
      {bool utf16 = false, Endian endian = Endian.big}) async {
    Uint8List decData = decryptDataFromFileSync(srcFilePath);
    String srcString;
    if ((decData[0] == 0xFE && decData[1] == 0xFF) ||
        (decData[0] == 0xFF && decData[1] == 0xFE)) {
      srcString = decData.toUtf16String();
    } else if (decData[0] == 0xEF && decData[1] == 0xBB && decData[2] == 0xBF) {
      srcString = decData.toUtf8String();
    } else {
      srcString =
          utf16 ? decData.toUtf16String(endian) : decData.toUtf8String();
    }
    return srcString;
  }

  /// Decrypts [srcFilePath] file to [destFilePath] file synchronously.
  ///
  /// If the argument [destFilePath] is not specified, decrypted file name is created
  /// by removing '.aes' file extension from [srcFilePath].
  /// If it has no '.aes' extension, then decrypted file name is created by adding
  /// '.decrypted' file extension to [srcFilePath].
  ///
  /// If decrypted file exists, the behaviour depends on [AesCryptOwMode].
  ///
  /// Returns [String] object containing the path to decrypted file.
  String decryptFileSync(String srcFilePath, [String destFilePath = '']) {
    srcFilePath = srcFilePath.trim();
    destFilePath = destFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        srcFilePath, 'Empty source file path.');
    if (srcFilePath == destFilePath) {
      throw AesCryptArgumentError(
          'Source file path and decrypted file path are the same.');
    }
    return _Cryptor.init(_passBytes, _owMode, _userdata)
        .decryptFileSync(srcFilePath, destFilePath);
  }

  /// Decrypts [srcFilePath] file to [destFilePath] file asynchronously.
  ///
  /// If the argument [destFilePath] is not specified, decrypted file name is created
  /// by removing '.aes' file extension from [srcFilePath].
  /// If it has no '.aes' extension, then decrypted file name is created by adding
  /// '.decrypted' file extension to [srcFilePath].
  ///
  /// If decrypted file exists, the behaviour depends on [AesCryptOwMode].
  ///
  /// Returns [Future<String>] that completes with the path to decrypted file
  /// once the entire operation has completed.
  Future<String> decryptFile(String srcFilePath,
      [String destFilePath = '']) async {
    srcFilePath = srcFilePath.trim();
    destFilePath = destFilePath.trim();
    AesCryptArgumentError.checkNullOrEmpty(_password, 'Empty password.');
    AesCryptArgumentError.checkNullOrEmpty(
        srcFilePath, 'Empty source file path.');
    if (srcFilePath == destFilePath) {
      throw AesCryptArgumentError(
          'Source file path and decrypted file path are the same.');
    }
    return await _Cryptor.init(_passBytes, _owMode, _userdata)
        .decryptFile(srcFilePath, destFilePath);
  }

//****************************************************************************
//**************************** CRYPTO FUNCTIONS ******************************
//****************************************************************************

  /// Creates random encryption key of [length] bytes long.
  ///
  /// Returns [Uint8List] object containing created key.
  Uint8List createKey([int length = 32]) => _Cryptor().createKey(length);

  /// Creates random initialization vector.
  ///
  /// Returns [Uint8List] object containing created initialization vector.
  Uint8List createIV() => _Cryptor().createKey(16);

  /// Computes SHA256 hash for binary data [data].
  ///
  /// Returns [Uint8List] object containing computed hash.
  Uint8List sha256(Uint8List data) => _Cryptor().sha256(data);

  /// Computes HMAC-SHA256 code for binary data [data] using cryptographic key [key].
  ///
  /// Returns [Uint8List] object containing computed code.
  Uint8List hmacSha256(Uint8List key, Uint8List data) =>
      _Cryptor().hmacSha256(key, data);

  /// Sets AES encryption key [key] and the initialization vector [iv].
  void aesSetKeys(Uint8List key, Uint8List iv) => _aes.aesSetKeys(key, iv);

  /// Sets AES mode of operation as [mode].
  ///
  /// Available modes:
  /// - [AesMode.ecb] - ECB (Electronic Code Book)
  /// - [AesMode.cbc] - CBC (Cipher Block Chaining)
  /// - [AesMode.cfb] - CFB (Cipher Feedback)
  /// - [AesMode.ofb] - OFB (Output Feedback)
  void aesSetMode(AesMode mode) => _aes.aesSetMode(mode);

  /// Sets AES encryption key [key], the initialization vector [iv] and AES mode [mode].
  void aesSetParams(Uint8List key, Uint8List iv, AesMode mode) {
    aesSetKeys(key, iv);
    aesSetMode(mode);
  }

  /// Encrypts binary data [data] with AES algorithm.
  ///
  /// Returns [Uint8List] object containing encrypted data.
  Uint8List aesEncrypt(Uint8List data) => _aes.aesEncrypt(data);

  // Decrypts binary data [data] encrypted with AES algorithm.
  //
  // Returns [Uint8List] object containing decrypted data.
  Uint8List aesDecrypt(Uint8List data) => _aes.aesDecrypt(data);
}
